# <div style="background-color:rgba(204, 229, 255, 0.5); text-align:center; vertical-align: middle; padding:40px 0; margin-top:30px"><span style="color:rgba(0, 76, 153, 1);">PHYS 121 Pre-Lab #4</span></div>
# Electric & Hydraulic Circuits – Week 1

***
## Learning Objectives:
* <b><span style="color:rgba(0, 153, 76, 1);"> Learn how to perform simple operations on DataFrames and arrays in Python.</span></b>

***

Over the next two weeks, you will work with both electric and hydraulic circuits.  You will see that concepts developed to analyze electrical circuits can also be applied to their hydraulic analogues.  In addition, you will attempt to uncover differences in the behaviour of the two systems and thus identify limitations of the analogy.) 

In the process, you will be collecting and manipulating data.  As before, a lot of code will be provided for you, but there will be instances where you will have to write your own lines of code.  This Pre-Lab is intended to provide a brief review of the kinds of operations you can use in Python and an introduction to how these operations can be applied to DataFrames and/or arrays.

**Basic Operations**

Recall the basic arithmetic operations in Python:

In [1]:
# Addition / Subtraction
a = 4 + 5
b = 2 - 3
print(a, b)

9 -1


In [2]:
# Division / Miltiplication
x = 10.2 * -4.0
y = 2 / 10
print(x, y)

-40.8 0.2


In [3]:
# Exponentiation
s = 12**2
print(s)

144


With numpy, we aquire a larger set of operations as well as a host of constants to work with. The notation is straight forward:

In [4]:
import numpy as np

# Pi and Euler's number e can be obtained as follows:
print(np.pi)
print(np.e)

3.141592653589793
2.718281828459045


In [5]:
# Natural logarithm
a = np.log(1)
b = np.log(np.e)

print(a, b)

0.0 1.0


In [6]:
# Trigonometric functions
c = np.cos(0)
d = np.sin(np.pi)
e = np.tan(np.pi / 2)

print(c, d, e)

1.0 1.2246467991473532e-16 1.633123935319537e+16


Be cautious when using these functions: they expect the input to be in units of radians. If you have a measurement in degrees, you can convert it to radians in the following way:

In [7]:
np.sin(np.radians(180))

1.2246467991473532e-16

You should also notice the rounding errors in the above. We should have $\sin\pi = 0$ and $\tan(\pi / 2) = \infty$, but instead we get something like $1.22\times10^{-16}$ and $1.63\times10^{16}$. These kinds of errors occur all the time in floating point arithmetic, so we sometimes have to be careful about what we're doing. [This article](https://docs.python.org/3/tutorial/floatingpoint.html) covers the phenomenon in more detail, but you shouldn't have to worry too much about it for this lab course. 

**Operating on Arrays and DataFrames**

Now, suppose we wanted to perform an operation on a set of data. All of the operations discussed above can be performed with entire numpy arrays or columns from DataFrames. For instance, consider pairwise addition of two datasets:

In [24]:
# When we add two arrays together, the output is another array
array1 = np.array([1, 2, 3])
array2 = np.array([4, 5, 0])

array3 = array1 + array2

print(array3)

[5 7 3]


In [36]:
# When we add two columns together, the output is an object called a "series", 
# which can be used to create an additional column.
import pandas as pd
df = pd.DataFrame({"c1" : [1, 2, 3],
                   "c2" : [4, 5, 0],
                  })

df["c3"] = df["c1"] + df["c2"]

print(df)

   c1  c2  c3
0   1   4   5
1   2   5   7
2   3   0   3


In [31]:
# If we try to do the same thing with lists, however... 
# Python will interpret the operation as "concatenation"
list1 = [1, 2, 3]
list2 = [4, 5, 0]

list3 = list1 + list2

print(list3)

[1, 2, 3, 4, 5, 0]


If using certain functions from the numpy package, you can apply them to lists as well as DataFrames and numpy arrays. However, multiplication does NOT work on lists.

In [46]:
# Trig functions act on each element individually
array4 = np.sin(array1)
print(array4)

print()

# When using the sine function from the numpy package, acting on a list will create a numpy array:
newArray = np.sin(list1)
print(newArray)
print(type(newArray))

[0.84147098 0.90929743 0.14112001]

[0.84147098 0.90929743 0.14112001]
<class 'numpy.ndarray'>


In [47]:
# Multiplication of arrays or DataFrame columns is performed pairwise
df["c4"] = df["c2"] * df["c1"]
print(df)

   c1  c2  c3  c4
0   1   4   5   4
1   2   5   7  10
2   3   0   3   0


In [48]:
# Multiplication is not defined for lists:
list1 * list2

TypeError: can't multiply sequence by non-int of type 'list'

The following section will guide you through the analysis of some experimental data, giving you an opportunity to practice performing operations on DataFrames. First, a bit of background on optics...

**Snell's law** describes the change in direction of light as it passes from one medium to another, called **refraction**. Refer to Fig 1. The incoming light strikes the interface at an angle of $\theta_1$ from perpendicular (the **angle of incidence**), and the outgoing light leaves at an angle $\theta_2$ (the **angle of refraction**). The relationship between these angles can be described in terms of the **refractive index** of the materials in question, denoted $n_1$ and $n_2$. Then Snell's law states that

\begin{align}
\frac{\sin\theta_1}{\sin\theta_2} = \frac{n_2}{n_1}
\label{eq:snell} \tag{1}
\end{align}

so in the figure below, since we can see that $\sin\theta_1 > \sin\theta_2$, medium two must have a larger refractive index than medium one, i.e. $n_2 > n_1$.

<p>
<center>
<img src="snell.svg" alt="image info" style="width: 35%; height: auto;" /><br>
<b>Fig. 1: The refraction of light passing from one medium to another. The incoming light strikes the interface at an angle of $\theta_1$ from perpendicular, and the outgoing light leaves at an angle $\theta_2$. The relationship between these angles and the refractive index of the materials is given by Snell's law. </b></center>
</p>

If the first medium is just air, these angles can be measured by shining a beam towards a hemispherical dish made of (or filled with) the second medium, as in Fig. 2. If the beam strikes the centre of the dish, no matter what the angle of refraction $\theta_2$ is, the beam will exit the dish perpendicular to its surface, so it will not be refracted.

<p>
<center>
<img src="snell2.svg" alt="image info" style="width: 45%; height: auto;" /><br>
<b>Fig. 2: The angles $\theta_1$ and $\theta_2$ can be measured using a hemispherical dish. </b></center>
</p>

The refractivity of a meterial can tell you about its composition. For example, the refractive index of seawater will increase as its salinity increases. The code cell below imports some generated "experimental" data on a sample of seawater held at $20^\circ\rm C$. We are going to analyze this data to determine the index of refraction of this seawater, which will tell us the salinity of the water. 

In [50]:
data = pd.read_csv("SampleData.csv")
print(data)

    theta1     theta2
0        5   3.724728
1       10   7.439369
2       15  11.147473
3       20  14.801432
4       25  18.375343
5       30  21.916742
6       35  25.352537
7       40  28.683443
8       45  31.861058
9       50  34.872212
10      55  37.693792
11      60  40.285501
12      65  42.550484
13      70  44.537203
14      75  46.138771
15      80  47.315403
16      85  48.029856


***
**<span style="color:blue">Question 1:</span>**  

First, we will calculate the sin of the given angles and add these as columns to our DataFrame. The cell block below does this for $\theta_1$. Repeat the process for $\theta_2$ in the empty code block after it. 

In [51]:
# Remember that we must also convert the angles to radians.
data["sin1"] = np.sin(np.radians(data["theta1"]))

***
**<span style="color:blue">Answer 1:</span>**

In [52]:
# Repeat the process above for the angle of refraction
data["sin2"] = np.sin(np.radians(data["theta2"]))
print(data)

    theta1     theta2      sin1      sin2
0        5   3.724728  0.087156  0.064963
1       10   7.439369  0.173648  0.129477
2       15  11.147473  0.258819  0.193335
3       20  14.801432  0.342020  0.255470
4       25  18.375343  0.422618  0.315241
5       30  21.916742  0.500000  0.373259
6       35  25.352537  0.573576  0.428187
7       40  28.683443  0.642788  0.479970
8       45  31.861058  0.707107  0.527861
9       50  34.872212  0.766044  0.571748
10      55  37.693792  0.819152  0.611441
11      60  40.285501  0.866025  0.646597
12      65  42.550484  0.906308  0.676240
13      70  44.537203  0.939693  0.701372
14      75  46.138771  0.965926  0.721020
15      80  47.315403  0.984808  0.735097
16      85  48.029856  0.996195  0.743493


***
**<span style="color:blue">Question 2:</span>**  

Now, if we calculate the ratio of these angles, we can apply Snell's law to determine the index of refraction of the water using Eq. 1. Using the code block below, create a new column in the DataFrame with this ratio. 

***
**<span style="color:blue">Answer 2:</span>**

In [11]:
data["ratio"] = data["sin1"] / data["sin2"]
print(data)

    theta1     theta2      sin1      sin2     ratio
0        5   3.724728  0.087156  0.064963  1.341621
1       10   7.439369  0.173648  0.129477  1.341151
2       15  11.147473  0.258819  0.193335  1.338708
3       20  14.801432  0.342020  0.255470  1.338788
4       25  18.375343  0.422618  0.315241  1.340621
5       30  21.916742  0.500000  0.373259  1.339553
6       35  25.352537  0.573576  0.428187  1.339548
7       40  28.683443  0.642788  0.479970  1.339225
8       45  31.861058  0.707107  0.527861  1.339570
9       50  34.872212  0.766044  0.571748  1.339829
10      55  37.693792  0.819152  0.611441  1.339707
11      60  40.285501  0.866025  0.646597  1.339359
12      65  42.550484  0.906308  0.676240  1.340217
13      70  44.537203  0.939693  0.701372  1.339792
14      75  46.138771  0.965926  0.721020  1.339665
15      80  47.315403  0.984808  0.735097  1.339698
16      85  48.029856  0.996195  0.743493  1.339884


***
**<span style="color:blue">Question 3:</span>**  

By definition, the refractive index of vacuum is 1. The refractive index of air is very close to 1, at about 1.000227. Using this and Eq. 1, create a new column in our data frame with the measured value of the refractive index $n_2$ of the seawater.

***
**<span style="color:blue">Answer 3:</span>**

In [12]:
data["n2"] = data["ratio"] * 1.000227
print(data["n2"])

0     1.341926
1     1.341456
2     1.339012
3     1.339092
4     1.340925
5     1.339857
6     1.339852
7     1.339529
8     1.339874
9     1.340133
10    1.340011
11    1.339663
12    1.340521
13    1.340096
14    1.339970
15    1.340002
16    1.340188
Name: n2, dtype: float64


We can take the average value and standard error in the mean for this data directly with Pandas. The syntax for this is 

```python
df["column"].mean()
df["column"].sem()
```

for the mean and standard error, respectively. 

***
**<span style="color:blue">Question 4:</span>**  

Use the code block below to calculate the mean and standard error for the refractive index of our seawater.

***
**<span style="color:blue">Answer 4:</span>**

In [13]:
avg = data["n2"].mean()
sem = data["n2"].sem()

print(f"n_2 = {avg:.4f} +/- {sem:.4f}")

n_2 = 1.3401 +/- 0.0002


In a 1995 paper, [Quan & Fry](https://opg.optica.org/ao/fulltext.cfm?uri=ao-34-18-3477&id=45728) provide an empirical equation for the refractivity of seawater as a function of salinity, temperature, and the wavelength of light. Suppose the wavelength of light used was $600 \rm nm$ and the temperature was measured to be $10 \rm ^\circ C$. Then Quan & Fry's equation reduces to 

\begin{align}
n(S) =  0.000188 S + 1.333355
\label{eq:refractivity} \tag{2}
\end{align}

where $S$ is the salinity in grams per litre. If we invert this expression, we obtain

\begin{align}
S(n) = 5312.08 n  - 7082.90
\label{eq:salinity} \tag{3}
\end{align}

with uncertainty

\begin{align}
\sigma_S = 5312.08 \sigma_n
\label{eq:salinityerr} \tag{4}
\end{align}


***
**<span style="color:blue">Question 5:</span>**  

Using the code block below, calculate the salinity of the sample and its uncertainty.

***
**<span style="color:blue">Answer 5:</span>**  

In [15]:
S = -7082.90 + 5312.08 * avg
errS = 5312.08 * sem

print(f"S = ({S} +/- {errS}) g/L")

S = (35.94498216364718 +/- 0.9637389771690847) g/L


And that's all there is to it! Now that you understand the basics, you'll be able to apply them in this week's lab.